core/controllers/profile.py

```diff
@@ -256,6 +261,17 @@ class ProfilePictureHandlerByUsernameHandler(base.BaseHandler):
 
     GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
 
+    URL_PATH_ARGS_SCHEMAS = {
+        'username': {
+            'schema': {
+                'type': 'basestring'
+            }
```

Comment:
Can we introduce more rules for username? I think there exist a limit for minimal/maximal length of the username and also the allowed characters.

Reply:
Hi, @vojtechjelinek after searching a lot into the project I got these validations.
1. Max length validation of 30 char
2. Only alphanumeric ['^[A-Za-z0-9]+$'] characters are allowed.
3. Should not be system reserved user name like oppia, admin.
![Screenshot from 2021-08-31 01-03-35](https://user-images.githubusercontent.com/42772952/131394767-2ead83b4-4038-42c2-92dd-9cadc0cf9f1d.png)

Reply:
@vojtechjelinek I think we need to add validators right ?

Reply:
@vojtechjelinek In other places we're using same username schema we need to add rules everywhere if we'll be updating it here.

Reply:
Yep

Reply:
@vojtechjelinek username validation is there when signing up and if in URL invalid username is passed no data will be returned.

Reply:
True, but I think at least the basic validation of length and allowed chars (regex) should be done here too.

Reply:
Hi, @vojtechjelinek I have added this method locally in `schema_utils.py`.

![Screenshot from 2021-08-31 16-44-36](https://user-images.githubusercontent.com/42772952/131493125-1d3b695a-99d6-47bd-b857-e5d485be5e96.png)


and in `profile.py`.

![Screenshot from 2021-08-31 16-43-58](https://user-images.githubusercontent.com/42772952/131493063-6f0a03ae-f205-4dd1-a80b-f6150f5c1efe.png)

Let me know if it is correct or not and what should I do next.

Reply:
@vojtechjelinek There is also repetition of code more importantly the **last part of checking system reserved name** can we add this validation somewhere else so that we can resuse it.

Reply:
Hmm, I might have some ideas, feel free to push the code so that I can give you comments into the code itself.

Reply:
@vojtechjelinek okay

Reply:
@vojtechjelinek just pushed changes, have a look and let me know what I can do.

Reply:
Hey, thinking about it I think we can import user_domain.py inside the schema_utils.py and then use the `user_domain.UserSettings.require_valid_username` in a try-except block, when the error is catched return `False` otherwise `True`. That way we can get rid of the duplication.

Reply:
Ya we can do that, I'll try and push it

Reply:
I am getting this error during git push

```
Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started
09:51:43 ERROR Third Party Python Lint Checks Manager: 6.8 secs

Unable to run the complete lint test, please check the following stack trace and fix the errors:
+--------------------------+
Traceback (most recent call last):
  File "/home/maahi/open-source/oppia/scripts/concurrent_task_utils.py", line 101, in run
    self.task_results = self.func()
  File "/home/maahi/open-source/oppia/scripts/linters/python_linter.py", line 255, in perform_all_lint_checks
    linter_stdout.append(self.lint_py_files())
  File "/home/maahi/open-source/oppia/scripts/linters/python_linter.py", line 123, in lint_py_files
    exit=False
  File "/home/maahi/open-source/oppia/../oppia_tools/pylint-quotes-0.1.9/pylint/lint/run.py", line 330, in __init__
    linter.load_plugin_modules(plugins)
  File "/home/maahi/open-source/oppia/../oppia_tools/pylint-quotes-0.1.9/pylint/lint/pylinter.py", line 540, in load_plugin_modules
    module = astroid.modutils.load_module_from_name(modname)
  File "/home/maahi/open-source/oppia/../oppia_tools/pylint-quotes-0.1.9/astroid/modutils.py", line 216, in load_module_from_name
    return importlib.import_module(dotted_name)
  File "/home/maahi/.pyenv/versions/3.7.10/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/maahi/open-source/oppia/scripts/linters/pylint_extensions.py", line 30, in <module>
    from core.controllers import payload_validator
  File "/home/maahi/open-source/oppia/core/controllers/payload_validator.py", line 25, in <module>
    import schema_utils
  File "/home/maahi/open-source/oppia/schema_utils.py", line 36, in <module>
    from core.domain.user_domain import UserSettings
  File "/home/maahi/open-source/oppia/core/domain/user_domain.py", line 31, in <module>
    (user_models,) = models.Registry.import_models([models.NAMES.user])
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 423, in import_models
    return cls._get().import_models(model_names)
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 187, in import_models
    from core.storage.user import gae_models as user_models
  File "/home/maahi/open-source/oppia/core/storage/user/gae_models.py", line 38, in <module>
    (base_models,) = models.Registry.import_models([models.NAMES.base_model])
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 423, in import_models
    return cls._get().import_models(model_names)
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 116, in import_models
    from core.storage.base_model import gae_models as base_model
  File "/home/maahi/open-source/oppia/core/storage/base_model/gae_models.py", line 52, in <module>
    transaction_services = models.Registry.import_transaction_services()
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 476, in import_transaction_services
    return cls._get().import_transaction_services()
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 249, in import_transaction_services
    from core.platform.transactions import cloud_transaction_services
  File "/home/maahi/open-source/oppia/core/platform/transactions/cloud_transaction_services.py", line 28, in <module>
    CLIENT = datastore.Client()
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/datastore/client.py", line 288, in __init__
    _http=_http,
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/client.py", line 277, in __init__
    _ClientProjectMixin.__init__(self, project=project, credentials=credentials)
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/client.py", line 225, in __init__
    project = self._determine_default(project)
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/datastore/client.py", line 317, in _determine_default
    return _determine_default_project(project)
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/datastore/client.py", line 93, in _determine_default_project
    project = _base_default_project(project=project)
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/_helpers.py", line 186, in _determine_default_project
    _, project = google.auth.default()
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/auth/_default.py", line 488, in default
    raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started

--------------------------------------------------

--------------------------------------------------
Some of the linting functions may not run until the above errors gets fixed
----------------------------------------
Please fix the errors below:
----------------------------------------
---------------------------
Checks Not Passed.
---------------------------
Push failed, please correct the linting issues above.
Already on 'handler-class-argument-schema'```

Reply:
Just remove the `(user_models,) = models.Registry.import_models([models.NAMES.user])` from _user_domain.py_ as it is not needed there.

Reply:
It worked Thanks.

Reply:
I have pushed changes please let me know what I should do next.